helix-term/src/commands.rs

```diff
@@ -3806,23 +3812,24 @@ fn remove_primary_selection(cx: &mut Context) {
 }
 
 pub fn completion(cx: &mut Context) {
+    // TODO completion starts to get ugly...
+    // maybe think about something like completion provider and separate completion-state into helix-view?
+    let clear_completion = async move {
+        let call: job::Callback =
+            Box::new(move |editor: &mut Editor, compositor: &mut Compositor| {
+                let ui = compositor.find::<ui::EditorView>().unwrap();
+                ui.clear_completion(editor);
+            });
+        Ok(call)
+    };
+    cx.jobs.callback(clear_completion);
```

Comment:
You're right, this is an ugly and potentially erroneous piece of code (as the comment above that suggests) and relies on the assumption that the pool handles the `clear_completion` first (because it doesn't need a round trip to the LSP).

This is done because it needs access to the `compositor` (for the completion menu) which is hold in the `Application`. And I'm not aware how to access the `Compositor` synchronously without having access to the `Application`.
I would really like to do this differently. But for this I think some more separation between the data and the view is necessary (in particular the completion menu).

Reply:
I see. Maybe we can use an `Arc<Mutex<bool>>` or a `std::sync::Once` and chain this future to the LSP requests with `and_then`? That way it's guaranteed to only run once and will always finish first.

Reply:
Well yeah why haven't I thought about that before... I've fixed that now with `helix_lsp::block_on(cx.jobs.finish())` between the clear completion and the other requests. I'm not sure though if it's easy to do this with chaining (the lsp requests have to run in parallel), because this way *all* the jobs are finished while blocking...